clients/lsp-javascript.el

```diff
@@ -790,20 +790,42 @@ name (e.g. `data' variable passed as `data' parameter)."
   (when-let ((workspace (lsp-find-workspace 'ts-ls (buffer-file-name))))
     (eq 'initialized (lsp--workspace-status workspace))))
 
-(defun lsp-clients-typescript-project-ts-server-path ()
-  "Return the project local TS server path."
-  (f-join (lsp-workspace-root) "node_modules" "typescript" "lib" "tsserver.js"))
+(defun lsp-clients-typescript-require-resolve (&optional dir)
+  "Get the location of the typescript.
+Use Node.js require.
+The node_modules directory structure is suspect
+and should be trusted as little as possible.
+If you call require in Node.js,
+it should take into account the various hooks.
+For example, yarn PnP.
+
+Optional argument DIR specifies the working directory
+to run the command in."
+  (let* ((default-directory (or dir default-directory))
+         (output
+          (string-trim-right
+           (shell-command-to-string
+            "node -e \"console.log(require.resolve('typescript'))\""))))
+    (if (string-empty-p output)
+        nil
+      (f-parent output))))
+
+(defun lsp-clients-typescript-package-path-direct ()
+  "`lsp-package-path' apply `'typescript' is modified.
+because the lsp server may not recognize the tsserver executable file,
+e.g., on Windows."
+  (if (memq system-type '(cygwin windows-nt ms-dos))
+      (lsp-clients-typescript-require-resolve (f-parent (lsp-package-path 'typescript)))
+    (lsp-package-path 'typescript)))
 
 (defun lsp-clients-typescript-server-path ()
   "Return the TS sever path base on settings."
-  (cond
-   ((and lsp-clients-typescript-prefer-use-project-ts-server
-         (f-exists? (lsp-clients-typescript-project-ts-server-path)))
-    (lsp-clients-typescript-project-ts-server-path))
-   (t
-    (if (memq system-type '(cygwin windows-nt ms-dos))
-        (f-join (f-parent (lsp-package-path 'typescript)) "node_modules" "typescript" "lib")
-      (f-join (f-parent (f-parent (lsp-package-path 'typescript))) "lib" "node_modules" "typescript" "lib")))))
+  (if lsp-clients-typescript-prefer-use-project-ts-server
+      (let ((server-path (lsp-clients-typescript-require-resolve)))
+        (if (f-exists? server-path)
+            server-path
+          (lsp-clients-typescript-package-path-direct)))
+    (lsp-clients-typescript-package-path-direct)))
```

Comment:
I think we can simplify this with `if-let*` block. 🤔 Try not to have `lsp-clients-typescript-package-path-direct` function in different places.

Reply:
I fixed.